src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/AbstractLanguageServer.cs

```diff
@@ -203,21 +218,93 @@ private void CheckServerState()
         }
     }
 
-    public async Task ShutdownAsync(string message = "Shutting down")
+    public async Task WaitForExitAsync()
+    {
+        lock (_lifeCycleLock)
+        {
+            // Ensure we've actually been asked to shutdown before waiting.
+            if (_shutdownRequestTask == null)
+            {
+                throw new InvalidOperationException("The language server has not yet been asked to shutdown.");
+            }
+        }
+
+        // Note - we wait for the _serverExitedSource task here instead of the _exitNotification task as we may not have
+        // finished processing the exit notification before a client calls into us asking to restart.
+        // This is because unlike shutdown, exit is a notification where clients do not need to wait for a response.
+        await _serverExitedSource.Task.ConfigureAwait(false);
+    }
+
+    /// <summary>
+    /// Tells the LSP server to stop handling any more incoming messages (other than exit).
+    /// Typically called from an LSP shutdown request.
+    /// </summary>
+    public Task ShutdownAsync(string message = "Shutting down")
+    {
+        Task shutdownTask;
+        lock (_lifeCycleLock)
+        {
+            // Run shutdown or return the already running shutdown request.
+            _shutdownRequestTask ??= Shutdown_NoLockAsync(message);
+            shutdownTask = _shutdownRequestTask;
+        }
+
+        return shutdownTask;
+    }
```

Comment:
```suggestion
        lock (_lifeCycleLock)
        {
            // Run shutdown or return the already running shutdown request.
            _shutdownRequestTask ??= Shutdown_NoLockAsync(message);
            return _shutdownRequestTask;
        }
    }
```

Reply:
is it ok that ShutdownAsync may be called multiple times?

Reply:
it can be called multiple times (as mentioned above due to multiple sources of shutdown).  But we only want to run once.